lib/src/widgets/adw/raw_target.dart

```diff
@@ -0,0 +1,156 @@
+  ///
+  /// Applies only to boxes with rectangular shapes; ignored if [shape] is not
+  /// [BoxShape.rectangle].
+  ///
+  /// {@macro flutter.painting.BoxDecoration.clip}
+  final BorderRadius borderRadius;
+
+  /// A border to draw above the background [color].
+  ///
+  /// Follows the [shape] and [borderRadius].
+  ///
+  /// Use [Border] objects to describe borders that do not depend on the reading
+  /// direction.
+  ///
+  /// Use [BoxBorder] objects to describe borders that should flip their left
+  /// and right edges based on whether the text is being read left-to-right or
+  /// right-to-left.
+  ///
+  final BoxBorder? border;
+
+  /// The shape to fill the background [color] into and
+  /// to cast as the [boxShadow].
+  ///
+  /// If this is [BoxShape.circle] then [borderRadius] is ignored.
+  ///
+  /// The [shape] cannot be interpolated; animating between two [BoxDecoration]s
+  /// with different [shape]s will result in a discontinuity in the rendering.
+  /// To interpolate between two shapes, consider using [ShapeDecoration] and
+  /// different [ShapeBorder]s; in particular, [CircleBorder] instead of
+  /// [BoxShape.circle] and [RoundedRectangleBorder] instead of
+  /// [BoxShape.rectangle].
+  ///
+  /// {@macro flutter.painting.BoxDecoration.clip}
+  final BoxShape shape;
+
+  /// A list of shadows cast by this box behind the box.
+  ///
+  /// The shadow follows the [shape] of the box.
+  final List<BoxShadow>? boxShadow;
+
+  /// The duration over which to animate the parameters of this container.
+  final Duration animationDuration;
+
+  /// The curve to apply when animating the parameters of this container.
+  final Curve animationCurve;
+
+  /// Status flag to denote that the widget is active.
+  /// Usually for popup buttons.
+  final bool isActive;
+
+  @override
+  State<StatefulWidget> createState() => _RawAdwTargetState();
+}
+
+class _RawAdwTargetState extends State<RawAdwTarget> {
+  late TargetStatus _status;
+
+  @override
+  void initState() {
+    super.initState();
+    _status = widget.isActive ? TargetStatus.active : TargetStatus.enabled;
+  }
+
+  @override
+  void didUpdateWidget(covariant RawAdwTarget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _status = widget.isActive ? TargetStatus.active : TargetStatus.enabled;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return MouseRegion(
+      cursor: SystemMouseCursors.click,
+      onEnter: (_) => setState(() => _status = TargetStatus.hover),
+      onExit: (_) => setState(
+        () => _status =
+            widget.isActive ? TargetStatus.active : TargetStatus.enabled,
+      ),
+      child: GestureDetector(
+        onTap: widget.onPressed,
+        onTapDown: (_) => setState(() => _status = TargetStatus.tapDown),
+        onTapUp: (_) => setState(
+          () => _status =
+              widget.isActive ? TargetStatus.active : TargetStatus.enabled,
```

Comment:
I think that it may be a good idea to separate the hover/tap logic.

For example, let's say I use the class with a pop-up (the comment line 101 seems to indicate that it's a possibility), and I have `_status = TargetStatus.active`. If I hover the pop-up, I will loose my active status. That does not seem correct to me, since the widget could be active _and_ hover? Am I missing something :S?

From my opinion, we should have two state elements: _statushover and _statusActive. This way I can hover over an active pop-up. Does that sound right to you :)?

Reply:
You're correct there. A `RawAdwTarget` derived widget cannot be active + hover at the same time, and it should be, because it's a real possibility.

I don't like the idea of having multiple variables to indicate a single thing like `_statusHover` or `_statusActive`. Having just a single variable, `_status` to denote the status of the widget is much more convenient when making the UI react according to the status.

The down side of using an enum here is that it cannot be two things at the same time. So I propose create a new status: `activeHover`, which will be activated when hovering an activated widget. This way, we can control that specific case, and not deal with multiple states that could be troublesome to handle, active, tap down and hovered statuses.

What do you think about this @MalcolmMielle ?

Reply:
It might be a matter of semantic but I think `active` should be the isolated case. `hover`, `tapUp`, and `tapDown` are three states that cannot coexist and fit in `_status`. On the other hand, `isActive` is an independent state if I understand correclty.

Looking at the class I'm actually confused as to why `widget.active` make sense here? It isn't part of the `RawAdwTarget` state, and only make it so that `_status` has value `TargetStatus.active` or `TargetStatus.enabled`. If it is used in pop-up, I would say that it belongs as a self contained state, in another `RawWidget` type (e.g. `RawActivableAdwTarget` that would inherit from `RawAdwTarget`) where the _status can be: hover -> clickDown -> clickUp and a clickDown would turn _stateActive to true (so that the pop-up is fired). This RawActivableAdwTarget could have a `show` abstract method that triggers the pop-up (see AdwComboRow for an example of the `show` method).

This way we would have `RawAdwTarget` handling `hover`, `tapDown`, `TapUp`, and a new class `RawActivableAdwTarget` handling `isActive` and the trigger of the method associated with `active`.

I know it's more work, but I feel like this would be much cleaner and the integration with `ComboRow` would then be straight forward :D.

Reply:
Ok I think I understand your explaination. Let's aim for one `_status` variable then. We still need to be able to hover+active at the same time in the `_status`. I've just ran a quick test in libadwaita demo and the headerbar button darken on hover even when active. The funny thing is that ComboRow on the other hand, does **not** have a hover color when active, but it's ok with your implementation since the callback `_resolveBackgroundColor` is class independent :). A good point for your implementation ;).

I think that, then the simplest solution is to increase the range of possible states to incorporate active VS inactive states. We would then have `enum TargetStatus { enabledHover, activeHover, tapDown }` as possible states, and all possible state would be stored in `_status`. This change means that we would need to update the logic in `MouseRegion` and `GestureDetector`, but that logic would be fully integrated in the `AdwButton` so that's great for the user.

At that point we need to differentiate between activable and non-activable button to not clutter the logic of the Button Widget (since we will use it for pop-up and "normal" buttons).
I would either:

* Create two class `AdwButton` and `ActivableAdwbutton`, where `AdwButton` does not have any logic associated with activability.
* Add a optional parameter to `AdwButton` called `activable` that default to `false` when the button isn't activable.

I would prefer option 1, because `ActivableAdwbutton` will need a required `onActive` callback as you said, and having it as an optional parameter for `AdwButton` might be awkward (?). However, I don't have a strong opinion on that, and I'm also fine with option 2 if you prefer.

What do you think?

Reply:
Great! I'll change the `TargetStatus` variable to include the new `activeHovered` status and rename the other hover state. Thank you very much for the feedback :)

Lastly, I'd prefer to bundle all into a complete widget. If we have to different buttons we'd need to copy and past all the `MouseRegion` and `GestureDetector` logic, which won't be ideal, or create a even more basic widget with that logic and build the status on top of that.

I think having a `onActive` parameter inside `AdwButton` is much better. If `null`, then the button will not be 'activable'.

What do you think? We're so close!

Reply:
Sounds good to me! Let's go :D

Reply:
Just added the new states and updated the logic. It works super great and mimics completely the behavior that features the GTK's libadwaita.

In the end, the `isActive` and all the logic related to that feature should be handled outside the widget, here why.
- The 'active' state can be invoke when the user clicks on a button to show a pop up. In this case, baking all the 'active' state directly to the button widget would work great.
- However, there are more use cases where buttons are active. If you check the header bar of the example app, you can see a drawer 'hide' button that is activated, meaning that the drawer is visible.

This last behavior is not compatible with baking the active state directly to the button. We need an external dependency to tell the widget whether it's active or not.

This behavior is normal in the rest of material buttons and widgets.

Let me know what you think about this, but I think this is the only solution, having an interface to communicate the 'active' status of the button from outside. @MalcolmMielle